api/src/main/java/io/strimzi/api/kafka/model/CustomResourceConditions.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.api.kafka.model;
+
+import io.fabric8.kubernetes.client.CustomResource;
+import io.strimzi.api.kafka.model.status.Status;
+
+import java.util.function.Predicate;
+
+public class CustomResourceConditions {
+
+    /**
+     * Returns a predicate that determines if a CustomResource is ready. A CustomResource is
+     * ready if the observedGeneration of its status is equal to the generation of its metadata
+     * and any of the conditions of its status has type:"Ready" and status:"True"
+     *
+     * @param <Y> the type of the custom resources Status
+     * @param <T> the type of the custom resource
+     * @return a predicate that checks if a CustomResource is ready
+     */
+    public static <Y extends Status, T extends CustomResource<?, Y>> Predicate<T> isReady() {
+        return isStatusLatestGenerationAndMatches(anyCondition("Ready", "True"));
+    }
+
+    /**
+     * Returns a predicate that determines if a KafkaRebalance is in a ProposalReady state. A KafkaRebalance is
+     * in proposal ready if the observedGeneration of its status is equal to the generation of its metadata
+     * and any of the conditions of its status has type:"ProposalReady" and status:"True"
+     *
+     * @return a predicate that checks if a KafkaRebalance is in ProposalReady state
+     */
+    public static Predicate<KafkaRebalance> isKafkaRebalanceProposalReady() {
```

Comment:
Why are we only exposing something for the ProposalReady state and not for all the states of `KafkaRebalanceState`?

Reply:
True, how about making the signature `isKafkaRebalanceState(KafkaRebalanceState state)` and using the enum name? Or if we moved it to KafkaRebalance, `isState`.

Also I was wondering if there should be any smarts around other terminal states, like you are waiting for `ProposalReady` but it enters `NotReady` instead due to an error. Then fabric8 would sit around until it hits the timeout.

Reply:
> True, how about making the signature `isKafkaRebalanceState(KafkaRebalanceState state)` and using the enum name? Or if we moved it to KafkaRebalance, `isState`.

Yeah, I think having a `state` parameter is a good idea. @scholzj how do you feel about putting these methods on the CR classes themselves?

> 
> Also I was wondering if there should be any smarts around other terminal states, like you are waiting for `ProposalReady` but it enters `NotReady` instead due to an error. Then fabric8 would sit around until it hits the timeout.

Doing that we would end up encoding assumptions about the current state machine within the predicate. Such assumptions could, _in general_, be invalidated by in the future by changes to the state macine. So I think we want to be a bit careful. 

The _specific_ suggestion you're making is: Should the predicate throw an exception if the SM enters a terminal state that's not the one for which the predicate will return true. The compatibility semantics of that depends on whether any states change their terminal-ness in the future. The Javadocs for `NotReady` suggest it's not a terminal state (can transition back to `New`). Overall I'm inclined to think it safest to let users handle this (e.g. by combining the predicates `ProposalReady && !NotReady`). But I wonder if @tomncooper @ppatierno have any thoughts about this.

Reply:
> Yeah, I think having a `state` parameter is a good idea. @scholzj how do you feel about putting these methods on the CR classes themselves?

I assume you mean classes such as `io.strimzi.api.kafka.model.Kafka`? I think that is fine with me. I don't have problem with that. I think the point you made about it being too generic was good.

Reply:
I will refactor to that